src/postgres-util/src/lib.rs

```diff
@@ -309,50 +317,88 @@ impl Config {
                 Ok(client)
             }
             SshTunnelConfig::Tunnel { .. } => {
+                let (session, local_port) = self.establish_openssh_connection().await?;
                 let tls = MakeTlsConnect::<TokioTcpStream>::make_tls_connect(&mut tls, &self.host)?;
-                let channel = self.establish_ssh_connection().await?;
+                let tcp_stream = TokioTcpStream::connect(("localhost", local_port)).await?;
                 let (client, connection) = self
                     .postgres_config
                     .replication_mode(ReplicationMode::Logical)
                     .connect_timeout(Duration::from_secs(30))
                     .keepalives_idle(Duration::from_secs(10 * 60))
-                    .connect_raw(channel, tls)
+                    .connect_raw(tcp_stream, tls)
                     .await?;
-                task::spawn(|| "postgres_connect_replication", connection);
+                task::spawn(|| "postgres_connect_replication", async {
+                    _ = connection.await;
+                    _ = session.close().await;
+                });
                 Ok(client)
             }
         }
     }
 
-    async fn establish_ssh_connection(
-        &self,
-    ) -> Result<AsyncChannel<TokioTcpStream>, anyhow::Error> {
+    async fn establish_openssh_connection(&self) -> Result<(openssh::Session, u16), anyhow::Error> {
         match &self.ssh_tunnel {
             SshTunnelConfig::Tunnel {
                 host,
                 port,
                 user,
-                public_key,
+                public_key: _,
                 private_key,
             } => {
-                let tcp_stream = TokioTcpStream::connect((host.as_str(), *port)).await?;
-                let mut session = AsyncSession::new(tcp_stream, Some(SessionConfiguration::new()))?;
-                session.handshake().await?;
+                let tempdir = tempfile::Builder::new()
+                    .prefix("ssh-tunnel-key")
+                    .tempdir()?;
+                let path = tempdir.path().join("key");
```

Comment:
can we just do this a `tempfile`?

Reply:
Unfortunately not, because with a `tempfile` I just get a handle to the open file, not a path I can pass to `openssh`. What I'm doing here is almost exactly what `tempfile` is doing under the hood, except that I keep the file alive just long enough to pass it to `openssh`.

Reply:
oh yeah, seems reasonable!